crates/containerd-shim-wasmedge/src/instance.rs

```diff
@@ -114,137 +106,45 @@ impl Instance for Wasi {
-                }
-                Err(Error::Others(e.to_string()))
-            }
-        }
+    fn get_id(&self) -> String {
+        self.id.clone()
     }
 
-    fn delete(&self) -> Result<(), Error> {
-        match instance_exists(&self.rootdir, self.id.as_str()) {
-            Ok(exists) => {
-                if !exists {
-                    return Ok(());
-                }
-            }
-            Err(err) => {
-                error!("could not find the container, skipping cleanup: {}", err);
-                return Ok(());
-            }
-        }
-        let container_root = get_instance_root(&self.rootdir, self.id.as_str())?;
-        let container = Container::load(container_root).with_context(|| {
-            format!(
-                "could not load state for container {id}",
-                id = self.id.as_str()
-            )
-        });
-        match container {
-            Ok(mut container) => container.delete(true).map_err(|err| {
-                Error::Any(anyhow!("failed to delete container {}: {}", self.id, err))
-            })?,
-            Err(err) => {
-                error!("could not find the container, skipping cleanup: {}", err);
-                return Ok(());
-            }
-        }
-
-        Ok(())
+    fn get_root_dir(&self) -> std::result::Result<PathBuf, Error> {
+        Ok(self.rootdir.clone())
     }
 
-    fn wait(&self, waiter: &Wait) -> Result<(), Error> {
-        let code = self.exit_code.clone();
-        waiter.set_up_exit_code_wait(code)
-    }
-}
+    fn build_container(&self) -> std::result::Result<Container, Error> {
+        fs::create_dir_all(&self.rootdir)?;
+        let stdin = maybe_open_stdio(self.stdin.as_str()).context("could not open stdin")?;
+        let stdout = maybe_open_stdio(self.stdout.as_str()).context("could not open stdout")?;
+        let stderr = maybe_open_stdio(self.stderr.as_str()).context("could not open stderr")?;
 
-impl Wasi {
-    fn build_container(
-        &self,
-        stdin: Option<i32>,
-        stdout: Option<i32>,
-        stderr: Option<i32>,
-    ) -> anyhow::Result<Container> {
         let syscall = create_syscall();
+        let err_others = |err| Error::Others(format!("failed to create container: {}", err));
         let container = ContainerBuilder::new(self.id.clone(), syscall.as_ref())
             .with_executor(vec![Box::new(WasmEdgeExecutor {
                 stdin,
                 stdout,
                 stderr,
-            })])?
-            .with_root_path(self.rootdir.clone())?
+            })])
+            .map_err(err_others)?
+            .with_root_path(self.rootdir.clone())
+            .map_err(err_others)?
             .as_init(&self.bundle)
             .with_systemd(false)
-            .build()?;
+            .build()
+            .map_err(err_others)?;
+        // Close the fds now that they have been passed to the container process
+        // so that we don't leak them.
+        stdin.map(close);
```

Comment:
I think we only care about `build` calling with an Ok result, but you remind me to close file descriptors on the wasmtime shim.

Reply:
I woiuld be nice if `maybe_open_stdio` returned an RAII wrapper.
A simple approach would be to return the `File` instead of the `RawFd`, and using `as_raw_fd()` when needed instead of using `into_raw_fd()` early on.

Reply:
> I woiuld be nice if `maybe_open_stdio` returned an RAII wrapper. A simple approach would be to return the `File` instead of the `RawFd`, and using `as_raw_fd()` when needed instead of using `into_raw_fd()` early on.

https://github.com/containerd/runwasi/pull/240 is similar to what you reference here, thought I am not sure what RAII is

Reply:
Sorry, that's my inner C++ betraying me.
It's basically when you have a struct implementing `Drop` to cleanup after itself.
When you `drop` a `File`, it automatically closes the file for you.
It is kind of what #240 does, but instead of using `into_raw_fd` the executors would hold a `Option<File>` instead of an `Option<RawFd>`, and the conversion to `RawFd` would be done at le last possible moment.

Reply:
I tried this but ran into some lifetime issues with the File and opted for a smaller change but if you have any suggestions be happy to fix that up 👍

Reply:
I was thinking something like this https://github.com/jsturtevant/runwasi/compare/use-cross-platfrom-file...jprendes:runwasi:use-cross-platfrom-file